File: ios/MusicPlayer.m
//
//  MusicPlayer.m
//  customIosApp
//
//  Created by Toni Kaplan on 9/27/19.
//  Copyright © 2019 Facebook. All rights reserved.
//

#import <Foundation/Foundation.h>

#import <React/RCTBridgeModule.h>
#import <React/RCTEventEmitter.h>


//@interface MusicPlayer : RCTEventEmitter <RCTBridgeModule>
@interface RCT_EXTERN_MODULE(MusicPlayer, RCTEventEmitter)

  RCT_EXTERN_METHOD(initalizeProgressTracker)
  RCT_EXTERN_METHOD(invalidateProgressTracker)

  RCT_EXTERN_METHOD(addPlayerStateObserver)

  RCT_EXTERN_METHOD(playGenre: (NSString *)genre)

  RCT_EXTERN_METHOD(getCurrentPlaybackTime: (RCTResponseSenderBlock)callback)
  RCT_EXTERN_METHOD(getCurrentSongDuration: (RCTResponseSenderBlock)callback)

@end

// React Native will not expose any function of MusicPlayer to React JavaScript unless explicitly done. To do so we have used RCT_EXPORT_METHOD() macro.

//@interface RCT_EXTERN_MODULE(MusicPlayer, RCTEventEmitter)

//RCT_EXTERN_METHOD(turnOn)
//
//RCT_EXTERN_METHOD(turnOff)
//
//RCT_EXTERN_METHOD(playGenre: (NSString *)genre)
//
//RCT_EXTERN_METHOD(getStatus: (RCTResponseSenderBlock)callback)
//
//RCT_EXTERN_METHOD(setOnProgress: (RCTResponseSenderBlock)callback)

//@end

File: ios/MusicPlayer.swift
//
//  MusicPlayer.swift
//  SpookyMusic
//
//  Created by Toni Kaplan on 9/27/19.
//  Copyright © 2019 Facebook. All rights reserved.
//

import Foundation
import MediaPlayer

//The @objc attribute makes your Swift API available in Objective-C and the Objective-C runtime.
@objc(MusicPlayer)class MusicPlayer: RCTEventEmitter {
  
  private var timer: Timer?
  
  override func supportedEvents() -> [String]! {
    return ["updateProgress", "updatePlayerState"]
  }
  
  @objc override static func requiresMainQueueSetup() -> Bool {
    return true
  }
  
  // Could alternatively use systemMusicPlayer to share the same music player as the music app
  @objc var musicPlayer = MPMusicPlayerController.applicationMusicPlayer
  
  @objc static var isMusicAuthorized = false;
  
  @objc func getCurrentPlaybackTime(_ callback: RCTResponseSenderBlock) {
    let currentPlaybackTime = self.musicPlayer.currentPlaybackTime as Double
    callback([NSNull(), currentPlaybackTime])
  }
  
  @objc func getCurrentSongDuration(_ callback: RCTResponseSenderBlock) {
    let nowPlayingItemDuration = self.musicPlayer.nowPlayingItem?.value(forProperty: "MPMediaItemPropertyPlaybackDuration") as! String
    callback([NSNull(), nowPlayingItemDuration])
  }
  
  @objc func pause(){
    musicPlayer.pause()
  }
  
  @objc func play(){
    musicPlayer.prepareToPlay()
    musicPlayer.play()
  }
  
  @objc func initalizePlayerWithPlaylist(_ playlist: String){
    let myPlaylistQuery = MPMediaQuery.playlists()
    musicPlayer.stop()
    let predicate = MPMediaPropertyPredicate(value: playlist, forProperty: MPMediaPlaylistPropertyName)
    myPlaylistQuery.addFilterPredicate(predicate)
    musicPlayer.setQueue(with: myPlaylistQuery)
    musicPlayer.prepareToPlay()
    musicPlayer.play()
  }
  
  // https://stackoverflow.com/questions/41755271/timer-doesnt-work-in-native-react-native-component
  @objc
  func initalizeProgressTracker(){
    self.timer = nil;
    self.invalidateProgressTracker();
    DispatchQueue.main.async(execute: {
      if(self.timer == nil){
        self.timer = Timer.scheduledTimer(
          withTimeInterval: 1,
          repeats: true,
          block: { timer in
            let nowPlayingItem = self.musicPlayer.nowPlayingItem
            if(nowPlayingItem != nil){
              let nowPlayingItemDuration = self.musicPlayer.nowPlayingItem?.value(forProperty: "playbackDuration") as! Double
              let currentTime = self.musicPlayer.currentPlaybackTime as Double
              let remainingTime = nowPlayingItemDuration - currentTime;
              
              let params: [String : Double]  = [
                "nowPlayingItemDuration": nowPlayingItemDuration,
                "currentTime": currentTime,
                "remainingTime": remainingTime
              ]
              
              self.sendEvent(withName: "updateProgress", body: params )
            }
            else{
              self.sendEvent(withName: "updateProgress", body: -1 )
            }
        }
        )
        
      }
    })
  }
  
  @objc func invalidateProgressTracker(){
    DispatchQueue.main.async(execute: {
      if self.timer != nil {
        self.timer!.invalidate()
        self.timer = nil
      }
    }
    )
  }
  
  @objc func updateNowPlayingInfo(){
    if(self.musicPlayer.nowPlayingItem != nil){
      let nowPlayingItemNameTemp = self.musicPlayer.nowPlayingItem?.value(forProperty: "title")
      let nowPlayingItemArtistTemp = self.musicPlayer.nowPlayingItem?.value(forProperty: "artist")
      let params: [String : String] = [
        "nowPlayingItemName": nowPlayingItemNameTemp as? String ?? "",
        "nowPlayingItemArtist": nowPlayingItemArtistTemp as? String ?? "",
      ]
      self.sendEvent(withName: "updatePlayerState", body: params )
    }
    else{
      let params : [String : String] = [
        "nowPlayingItemName": "",
        "nowPlayingItemArtist": "",
      ]
      self.sendEvent(withName: "updatePlayerState", body: params )
    }
  }
  
  @objc func addPlayerStateObserver(){
    musicPlayer.beginGeneratingPlaybackNotifications()
    
    NotificationCenter.default.addObserver(self, selector:#selector(self.updateNowPlayingInfo), name: NSNotification.Name.MPMusicPlayerControllerNowPlayingItemDidChange, object: nil)
    NotificationCenter.default.addObserver(self, selector:#selector(self.updateNowPlayingInfo), name: NSNotification.Name.MPMusicPlayerControllerPlaybackStateDidChange, object: nil)
  }
}
